Policy regeneration improvements

jrajahalme opened the pull request:
Restructure policy regeneration code to keep track if the policy actually changed or not and also restructure the proxy test to reuse the same containers throughout all 20 test variations.

Together these changes reduce the 10-proxy.sh runtime on a 6-CPU/8GB Linux 4.13 VM from 20 to 4.3 minutes.

tgraf commented on pkg/policy/consumer.go:
```diff
@@ -245,32 +245,37 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
 		c.addToMaps(id)
 		c.Consumers[id.StringID()] = NewConsumer(id)
+		return true
 	} else {
```
You can get rid of the else branch here now, go vet will complain about this later otherwise.

jrajahalme replied:
Done.

tgraf commented on daemon/policy.go:
```diff
@@ -93,6 +93,7 @@ func (d *Daemon) EnableEndpointPolicyEnforcement(e *endpoint.Endpoint) bool {
 	} else if policy.GetPolicyEnabled() == endpoint.DefaultEnforcement {
 		// Default mode means that if rules contain labels that match this endpoint,
 		// then enable policy enforcement for this endpoint.
+		// TODO: Could check e.Consumable instead? Would be much cheaper.
```
I've created a GH issue, change the comment to `GH-1676: Could check ...`

jrajahalme replied:
OK.

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -54,37 +50,15 @@ func (e *Endpoint) checkEgressAccess(owner Owner, opts models.ConfigurationMap,
 	if !e.Opts.IsEnabled(OptionConntrack) {
-		e.Consumable.AllowConsumerAndReverseLocked(cache, id)
+		return e.Consumable.AllowConsumerAndReverseLocked(cache, id)
 	} else {
```
You can remove the else branch here now, otherwise go vet will complain later.

jrajahalme replied:
OK.

aanm commented on daemon/daemon.go:
```diff
@@ -1095,6 +1095,7 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 
 	// Only recompile if configuration has changed.
 	if changes > 0 {
+		d.policy.BumpRevision() // force policy recalculation
```
Do we really need policy recalculation if the daemon's configuration is changed?

jrajahalme replied:
Changes in policy enforcement have on effect on how the endpoint specific bpf program is generated. I changed the code in endpoint/policy.go to only regenerate if the policy changes, and did not want complicate that by, e.g., revisioning the option changes.

ianvernon replied:
On line `1080-1082` we check if policy enforcement has been changed, and then trigger policy updates for the endpoint. This function call should eventually bump the policy revision, so I don't see a need for this addition.

aanm commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +100,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
+		}
+		// Skip currently unused IDs
+		if lbls == nil || len(lbls) == 0 {
+			continue
```
Out of curiosity, was this condition ever reached?

jrajahalme replied:
I don't know, maybe not likely for the reserved identities, but must be likely when looping from min to max ID as I don't think we compact the ID space after IDs become unused.

aanm commented on pkg/policy/l3.go:
```diff
@@ -133,6 +134,23 @@ func (m L3PolicyMap) DeepCopy() L3PolicyMap {
 }
 
+func ipnetEqual(a, b net.IPNet) bool {
+	return bytes.Equal(a.IP, b.IP) && bytes.Equal(a.Mask, b.Mask)
```
net.IP has the Equal method. `return a.IP.Equal(b.IP) && ...`

jrajahalme replied:
Documentation states: "An IPv4 address and that same address in IPv6 form are considered to be equal." I though that might not be desirable and decided to err on the safe side; Deeming policies in-equal cases recalculation and while inefficient, calculating the same policy again is not breaking anything functionally.

aanm commented on pkg/policy/l4.go:
```diff
@@ -35,6 +35,29 @@ const (
+}
+
+func kafkaEqual(a, b api.PortRuleKafka) bool {
+	if a.APIVersion != b.APIVersion || a.APIKey != b.APIKey || a.Topic != b.Topic {
```
Why not `return a.APIVersion == b.APIVersion && a.APIKey == b.APIKey && a.Topic == b.Topic`?

jrajahalme replied:
No reason other that copy paste from HTTP case that required looping through the header matches after. I changed this as you suggested :-)

aanm reviewed (commented):
Just small nits, overall looks good!

tgraf commented on pkg/endpoint/policy.go:
```diff
@@ -303,19 +322,23 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (bool, error) {
-	if !policyChanged {
+	if !(policyChanged || optsChanged) {
 		e.policyRevision = revision
+		e.nextPolicyRevision = revision
```
You can update `nextPolicyRevision` outside of the branch

jrajahalme replied:
Right!

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
-func (e *Endpoint) regenerateConsumable(owner Owner) (bool, error) {
-	c := e.Consumable
+func getLabelsMap(owner Owner, cache *policy.ConsumableCache, minID, maxID policy.NumericIdentity) (labelArrays, error) {
+	labelsMap := map[policy.NumericIdentity]labels.LabelArray{}
```
Use the `labelArrays` type you declared above here.

jrajahalme replied:
OK. Also responding to the previous comment here, as there is no response option there:

Daemon's TriggerPolicyUpdates() incremented policy cache's iteration number, but that was not used for anything, so the BumpRevision was actually needed here. I have further refactored the TriggerPolicyUpdates to Bump the repository revision instead, so that after that your comment applies. I also changed the Opts.Apply() changedOptions callback to also bump the repo revision so that the line that I added above is no longer needed even in that case. This based on assumption that some other daemon options may require policy regeneration as well, not 100% sure of that, though.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
-		return false, err
+	// Skip consumable recomputation for this consumable is already valid
+	if c.Iteration == revision {
+		log.Debugf("Reusing cached policy for identity %d", c.ID)
```
Can you use `log.WithFields` to refer to the identity? For example, from `daemon/endpoint.go`:
```
log.WithFields(log.Fields{
				logfields.EndpointID: ep.StringID(),
				logfields.Identity:   identity.ID,
			}).WithError(err).Warningf("Unable to release temporary identity")
```

tgraf replied:
FYI: We started using this recently, it leads to consistent grep'able log messages which is great. We should convert over log messages to this as we go along.

jrajahalme replied:
I'll need to do another pass at this.

jrajahalme replied:
Done.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
+		return false, nil
 	}
 
-	c.Mutex.RLock()
```
Why are we removing `RLock()` here? Should the consumable's mutex always be locked when this function is called? If so, update the description of this function to ensure that it specifies that the Consumable's lock should be held before calling this function.

jrajahalme replied:
I have since refactored this so that this should be clearer in the current version. The general problem is that we can't promote a read lock to a write lock without releasing the lock in between.

In this case (read) locking may have been mainly decorative as the consumable's labels should not change during the lifetime of the consumable, and the endpoint itself has a reference to it.

I've clarified the locking in the current revision of the series.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
-	if c.L4Policy != nil {
-		e.cleanUnusedRedirects(owner, c.L4Policy.Ingress, newL4policy.Ingress)
-		e.cleanUnusedRedirects(owner, c.L4Policy.Egress, newL4policy.Egress)
+	if !c.L4Policy.PolicyEqual(newL4policy) {
```
Add some comments here describing what is going on, would be helpful to someone who is new to Cilium.

jrajahalme replied:
I'll do an additional pass on commenting this. Thanks for the suggestion!

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
-				e.PolicyID(), err)
+	for srcID, srcLabels := range labelsMap {
+		ctx.From = srcLabels
+		log.Debugf("[%s] Evaluating context %+v", e.PolicyID(), ctx)
```
Similar to comment above, use `log.WithFields` for the ID of the endpoint.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
 
-// regeneratePolicy returns whether the policy for the given endpoint should be
-// regenerated. Only called when e.Consumable != nil.
+// regeneratePolicy regenerates endpoints policy if needed and returns
```
endpoints --> endpoint's

jrajahalme replied:
ok.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
+// Policy generation may fail, and in that case we exit before
+// actually changing the policy in any way, so that the last policy
+// remains fully in effect if the new policy can not be
+// implemented. This is done on endpoint-basis, however, and it is
```
... on a per-endpoint basis...

jrajahalme replied:
OK.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -126,172 +99,217 @@ func (e *Endpoint) cleanUnusedRedirects(owner Owner, oldMap policy.L4PolicyMap,
+	e.checkEgressAccess(owner, labelsMap[policy.ReservedIdentityWorld], opts, OptionAllowToWorld)
 
-	if e.Consumable != nil && e.Consumable.L4Policy.RequiresConntrack() {
+	if e.Consumable.L4Policy.RequiresConntrack() {
```
`e.Consumable` --> `c`

jrajahalme replied:
Thanks!

ianvernon commented on pkg/policy/consumer.go:
```diff
@@ -245,32 +245,36 @@ func (c *Consumable) removeFromMaps(id NumericIdentity) {
-		log.Warningf("Allowed a consumer %d->%d which can't be found in the reverse direction", c.ID, id)
+		return true
 	}
+	log.Warningf("Allowed a consumer %d->%d which can't be found in the reverse direction", c.ID, id)
```
As above, use `log.WithFields`.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -50,6 +70,31 @@ type L4Filter struct {
 }
 
+// PolicyEqual returns true if the L4 filters are the same.
+func (l4 L4Filter) PolicyEqual(l4b L4Filter) bool {
```
Add unit tests.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -105,6 +150,19 @@ func (l4 L4Filter) String() string {
 type L4PolicyMap map[string]L4Filter
 
+// PolicyEqual returns true if the L4 policy maps are the same.
+func (l4 L4PolicyMap) PolicyEqual(l4b L4PolicyMap) bool {
```
Add unit tests.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -222,3 +280,14 @@ func (l4 *L4Policy) DeepCopy() *L4Policy {
 }
+
+// PolicyEqual returns true if the L4 policies are the same.
+func (l4 *L4Policy) PolicyEqual(l4b *L4Policy) bool {
```
Add unit tests.

jrajahalme replied:
Changed to use reflect.DeepEqual

ianvernon commented on pkg/policy/l3.go:
```diff
@@ -155,6 +173,17 @@ func (l3 *L3Policy) DeepCopy() *L3Policy {
 }
 
+// PolicyEqual returns true if policy described by 'n' is equal to that of 'l3'.
+func (l3 *L3Policy) PolicyEqual(n *L3Policy) bool {
```
Add unit tests.

ianvernon commented on pkg/policy/l3.go:
```diff
@@ -133,6 +134,23 @@ func (m L3PolicyMap) DeepCopy() L3PolicyMap {
+
+// PolicyEqual returns true if policy described by 'n' is equal to that of 'm'.
+func (m L3PolicyMap) PolicyEqual(n L3PolicyMap) bool {
+	if m.IPv6Count != n.IPv6Count || m.IPv4Count != n.IPv4Count {
```
Add unit tests.

jrajahalme replied:
Replaced with reflect.DeepEqual

ianvernon commented on pkg/policy/l3.go:
```diff
@@ -133,6 +134,23 @@ func (m L3PolicyMap) DeepCopy() L3PolicyMap {
 }
 
+func ipnetEqual(a, b net.IPNet) bool {
+	return bytes.Equal(a.IP, b.IP) && bytes.Equal(a.Mask, b.Mask)
```
Add unit tests.

jrajahalme replied:
Changed to use reflect.DeepEqual instead.

rlenglet commented on daemon/endpoint.go:
```diff
@@ -224,57 +230,67 @@ func (h *patchEndpointID) Handle(params PatchEndpointIDParams) middleware.Respon
 	}
 
-	if epTemplate.Mac != "" {
+	if epTemplate.Mac != "" && !reflect.DeepEqual(ep.LXCMAC, newEp.LXCMAC) {
```
Just a nitpick. `LXCMAC` is a `mac.MAC`, which is defined as a `net.HardwareAddr`, which in turn is defined as a `[]byte`. So it should be possible to use `bytes.Compare` to compare the 2 values, which should be more efficient than `reflect.DeepEqual`.
Not a blocker.

jrajahalme replied:
Done, and figured that addresses are also byte slices so using bytes.Compare for them as well.

rlenglet commented on pkg/endpoint/bpf.go:
```diff
@@ -335,8 +337,44 @@ func (e *Endpoint) runInit(libdir, rundir, epdir, debug string) error {
+}
+
+// Must be called when endpoint is still locked.
+func (e *Endpoint) createEpFronted() *epFrontend {
```
Typo: `createEpFronte`__n__`d`.

jrajahalme replied:
Fixed.

rlenglet commented on pkg/endpoint/bpf.go:
```diff
@@ -345,15 +383,44 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
+		e.Mutex.Unlock()
+		log.WithFields(log.Fields{
+			logfields.EndpointID: e.StringID(),
+		}).Debugf("Skipping build due to invalid state: %s", e.state)
```
Why not pass the state as a log field?

jrajahalme replied:
Logging also fixed.

rlenglet commented on pkg/endpoint/bpf.go:
```diff
@@ -345,15 +383,44 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
+		// policy change.
+		if _, err = e.regeneratePolicy(owner, nil); err != nil {
+			e.Mutex.Unlock()
+			return fmt.Errorf("Unable to regenerate policy for '%s': %s",
```
Unable -> unable (for consistency)

jrajahalme replied:
Went the other way (consistently starting everything with a capital letter, dropping punctuation from the end).

rlenglet commented on pkg/endpoint/policy.go:
```diff
@@ -227,123 +170,143 @@ func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy,
+		err := e.applyL4PolicyLocked(labelsMap, e.L4Policy, c.L4Policy)
+		if err != nil {
+			// This should not happen, and we can't fail at this stage anyway.
+			log.Fatalf("L4 Policy application failed: %v", err)
```
`log.WithError(err).Fatal("L4 Policy application failed")`

rlenglet commented on pkg/endpoint/policy.go:
```diff
@@ -352,98 +315,174 @@ func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
 
 	if valid {
+		if reflect.DeepEqual(e.L3Policy, newL3policy) {
+			e.getLogger().Debugf("No change in CIDR policy.")
```
`e.getLogger().Debug("No change in CIDR policy")`

rlenglet commented on pkg/endpoint/policy.go:
```diff
@@ -352,98 +315,174 @@ func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
+	// through it each time.
+	labelsMap, err := getLabelsMap(owner)
 	if err != nil {
+		log.Debugf("[Endpoint %d] Received error while evaluating policy: %s", e.ID, err)
```
Use a log field for the endpoint ID?

nebril commented on pkg/endpoint/endpoint.go:
```diff
@@ -1263,40 +1225,139 @@ func (e *Endpoint) GetDockerNetworkID() string {
+// SetStateLocked modifies the endpoint's state
+// endpoint.Mutex must be held
+// Returns true only if endpoints state was changed as requested
+func (e *Endpoint) SetStateLocked(toState string) bool {
```
Would be nice to have this method unit tested to ensure that all transitions are correct. Unit test would also act as a documentation on what transitions are allowed.

jrajahalme replied:
Good point. I've added a full matrix as well as state transitions during a typical lifecycle of an endpoint.

nebril commented on pkg/endpoint/endpoint.go:
```diff
@@ -1263,40 +1225,139 @@ func (e *Endpoint) GetDockerNetworkID() string {
-	return e.IsDisconnectingLocked()
+// SetStateForTestingOnly modifies the endpoint's state
+// This is used when creating endpoints for testing, so no locking is done.
+func (e *Endpoint) SetStateForTestingOnly(toState string) {
```
This method breaks hiding the `Endpoint.state` field as private. Since it's used only for newly created endpoints in tests, I would suggest to add `NewEndpointWithState` function in `pkg/endpoint` which would accept string and return new endpoint struct with state set. Then it can be populated with all public fields needed for tests.

This may seem a bit over the top, but I think it would make the code better.

jrajahalme replied:
Endpoint initialization in unit tests became a bit uglier, but overall it's better so I did this too.

nebril commented on pkg/endpoint/policy.go:
```diff
@@ -352,98 +323,197 @@ func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
+	e.getLogger().Debug("Starting regenerate...")
+
+	// Collect label arrays before policy computation, as this can fail.
+	// GH-XXXX should allow optimizing this away, but currently we can't
```
Please fill XXXX

jrajahalme replied:
Done (GH-1128)

nebril commented on pkg/endpoint/policy.go:
```diff
@@ -352,98 +323,197 @@ func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
+	defer repo.Mutex.RUnlock()
+
+	// Recompute policy for this endpoint only if not already done for this revision.
+	if !e.forcePolicyCompute && e.nextPolicyRevision >= revision && labelsMap == e.LabelsMap && opts == nil {
```
Please break this line.

jrajahalme replied:
Done.

nebril commented on pkg/endpoint/policy.go:
```diff
@@ -562,50 +620,56 @@ func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
 			e.Consumable.Mutex.Unlock()
 			return
 		}
+		// XXX: This removes the consumable from the cache even if other endpoints
```
What's XXX?

jrajahalme replied:
Changed "XXX" to "TODO". My intention was to get reviewer's attention and maybe think it over and give some feedback :-)

nebril reviewed (changes requested):
Looks good overall, 2 code-related suggestions inline, please clean the comments.

tgraf reviewed (approved):
Fantastic work!

ianvernon commented on daemon/endpoint.go:
```diff
@@ -159,9 +160,18 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
+		ready = ep.SetStateLocked(endpoint.StateWaitingToRegenerate)
+	}
+	ep.Mutex.Unlock()
+	if ready {
```
What happens if line 166 (`if ep.GetStateLocked() == endpoint.StateReady`) returns `false`? The endpoint appears to not be regenerated in that case, but we return `NewPutEndpointIDCreated()`

jrajahalme replied:
My intention was not to change the behavior here, but upon checking this again I see that if the initial state was waiting-for-identity master would kick a build at this time, while this depended upon a following PATCH to do the initial build in that case. I've added a patch that adds a check for waiting-for-identity as well so that it behaves the same as master. Note that it is still possible that the initial state is "creating" (or "disconnected" in master) and in this case the initial build is (and was) no done at this stage, but when processing the ensuing PATCH request that changes the state to waiting-for-identity (or ready).

jrajahalme replied:
I'm not 100% sure what the implication of no bpf build is between the PUT and PATCH in the case the initial state is not waiting-for-identity or ready, but that is how the master behaves, so if we want to change that, I'd suggest doing it in a separate PR.

ianvernon replied:
Makes sense! Thank you for the clear explanation.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -227,123 +175,146 @@ func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy,
+			err := e.applyL4PolicyLocked(labelsMap, e.L4Policy, c.L4Policy)
+			if err != nil {
+				// This should not happen, and we can't fail at this stage anyway.
+				e.getLogger().Fatal("L4 Policy application failed")
```
Do we want to run `Fatal` here? This will crash the daemon?

jrajahalme replied:
This can fail only if the bpf map update fails, and that is an error we can not recover from, and it may be better to have cilium restart.

That said, I hope that in future we can push bpf map updates one layer down for policy computation.

ianvernon replied:
Thanks for clarifying, makes sense.

ianvernon commented on tests/16-cidr-ingress-policy.sh:
```diff
@@ -121,7 +120,7 @@ docker run --rm -i --net ${TEST_NET} -l "${ID_SERVICE2}" --cap-add NET_ADMIN ${D
 
 log "importing L3 CIDR policy for IPv4 egress"
-policy_delete_and_wait "--all"
+cilium policy delete --all
```
Why were calls to `policy_delete_and_wait "--all"` removed?

jrajahalme replied:
Functionally it is not necessary to wait after deletion as a new policy is being installed right after and we wait for that. This allows cilium to skip regenerating bpf for the deleted policy in some cases, depending on how fast the bpf rebuilds are scheduled after policy deletion.

ianvernon replied:
Makes sense, thanks.

ianvernon commented on tests/k8s/Vagrantfile:
```diff
@@ -89,7 +89,7 @@ Vagrant.configure(2) do |config|
 		vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
 		vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
-
+                vb.memory = 3072
```
:)

ianvernon reviewed (commented):
Still need to review more dense sections, have a few questions for now.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -305,10 +306,11 @@ func writeGeneve(prefix string, e *Endpoint) ([]byte, error) {
-	e.Mutex.RLock()
-	args := []string{libdir, rundir, epdir, e.IfName, debug}
+func (e *Endpoint) runInit(libdir, rundir, epdir, ifName, debug string) error {
+	args := []string{libdir, rundir, epdir, ifName, debug}
```
Is `e.IfName` ever different from the epFrontend.IfName`?

jrajahalme replied:
endpoint's IfName can be set with the API "PATCH /endpoint/{id}" request. Not sure if we support changing it after it is set for the first time, but I presume it could be un-set after the endpoint is created?

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -335,8 +337,44 @@ func (e *Endpoint) runInit(libdir, rundir, epdir, debug string) error {
 
+// Implement EndpointFrontend interface, holds values read from the endpoint from which the
+// bpf code and maps were generated; to be used after the lock is released.
+type epFrontend struct {
```
I found the `Frontend` name here and the explanation to be a bit confusing, do you think that the following would be clearer?

```
// epInfoCache describes the set of lxcmap entries necessary to describe an Endpoint
// in the BPF maps. It may be generated while holding the Endpoint lock, then used
// after releasing that lock to push the entries into the datapath.
type epInfoCache {
...
```

If I follow correctly, the main intention of this structure is to ensure that the lxcmap entries that we generate are consistent with the rest of the map entry generation for this Endpoint, which we guarantee by creating this cache while holding the lock. With these changes, after the lock is released, the Endpoint could be modified but that won't affect the ongoing generation of lxcmap entries.

(Separately I wonder if EndpointFrontend could be better named/described, but that's not introduced by this work and could be taken up separately..)

jrajahalme replied:
Right, we get to generate the bpf for the endpoint as if we held the ep mutex for the whole time by caching all the ep info we are going to use after bpf compilation while still holding the mutex.

I adopted your proposed comment and type name change, thanks!

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -335,8 +337,44 @@ func (e *Endpoint) runInit(libdir, rundir, epdir, debug string) error {
+
 // regenerateBPF rewrites all headers and updates all BPF maps to reflect the
 // specified endpoint.
+// Must be called with endpoint mutex not held.
```
Also, must be held with `endpoint.BuildMutex` held.

jrajahalme replied:
Right, I amended the comment accordingly.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -345,15 +383,36 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
+		if _, err = e.regeneratePolicy(owner, nil); err != nil {
+			e.Mutex.Unlock()
+			return fmt.Errorf("Unable to regenerate policy for '%s': %s",
+				e.PolicyMap.String(), err)
```
If `regeneratePolicy()` fails in `getLabelsMap()`, then it seems possible for `e.PolicyMap` to remain uninitialized, which we would then attempt to format as a string in this error creation, which would then lead to dereference of nil?

jrajahalme replied:
Thanks for noticing this, I removed e.PolicyMap from the error message in this case. This block is executed only in the dry mode and the e.PolicyMap is not supposed to exist in dry mode anyway!

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -352,98 +323,199 @@ func (e *Endpoint) regenerateL3Policy(owner Owner) (bool, error) {
+	// through it each time.
+	labelsMap, err := getLabelsMap(owner)
 	if err != nil {
+		e.getLogger().WithError(err).Debug("Received error while evaluating policy")
```
I think the callers will log this? Or are they just passing errors all the way back up and never logging them?

jrajahalme replied:
I have added some debugging, but IMO the equivalent log message existed also before. That said, the some callers log the error, but with less context, and others pass it to the endpoint's Status Log (reported with "cilium endpoint get") without logging it at all.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -406,65 +462,71 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
+		if err == nil {
+			// Mark the endpoint to be running the policy revision it was
+			// compiled for
+			e.bumpPolicyRevision(epFrontend.revision)
```
:thumbsup:

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -755,25 +725,9 @@ func (e *Endpoint) String() string {
-	e := data.(*Endpoint)
-	switch key {
-	case OptionConntrack:
-		e.invalidatePolicy()
```
Is this linked to the commandline options where you can enable/disable conntrack on a per-endpoint basis? Why can we safely remove this?

ianvernon replied:
+1 - what rationalizes removing this?

jrajahalme replied:
The need for this is replaced by passing options from API requests to regeneratePolicy() via the new opts argument. regeneratePolicy will then use this set of options as basis for it's own options manipulation and return true if bpf needs to be regenerated. Only regeneratePolicy() will call ApplyOptsLocked() so the side effects of the optionsChanged() callback are not needed any more.

In the case where API disables conntrack for the endpoint and the applicable policy requires conntrack, this option change will be reverted by regeneratePolicy(). Via this refactoring we do not invalidate endpoint's policy in this case and avoid bpf regeneration.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -18,86 +18,50 @@ import (
+func (e *Endpoint) checkEgressAccess(owner Owner, dstLabels labels.LabelArray, opts models.ConfigurationMap, opt string) {
 	ctx := policy.SearchContext{
 		From: e.Consumable.LabelArray,
+		To:   dstLabels,
```
I was a bit surprised that this wasn't called out specifically in the patch description or indeed a separate patch, so it seems like this was a bug in that previously egress label-based access couldn't be properly applied with to:host or to:world restrictions. Are there tests for this?

jrajahalme replied:
ctx.To was assigned separately, see some of the deleted lines below. There should be no functional change here.

joestringer commented on pkg/endpoint/policy.go:
```diff
@@ -227,123 +175,146 @@ func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy,
+			}
+		}
+		e.L4Policy = c.L4Policy // Reuse the common policy
+		e.LabelsMap = labelsMap // Remember the set of labels used
```
Should drymode change these?

jrajahalme replied:
My understanding is that the dry mode should do everything that can be done without root (or net admin) privileges. The intention is to enable unit testing of the implementation, so the more "real" it is the better.

joestringer reviewed (commented):
I looked at the BPF changes, seems to be mostly removing the `DeepCopy()` implementations which as far as I can tell are currently unused. I also provided some comments from a policy perspective, where there might be one or two unintended state modifications or nil dereferences.

nebril commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -241,15 +163,119 @@ func (s *EndpointSuite) TestEndpointStatus(c *C) {
-		ID: 12,
-	}
-	c.Assert(ep.PolicyID(), Equals, "Policy ID 0")
+func (s *EndpointSuite) TestEndpointState(c *C) {
```
Thanks for adding this!

nebril reviewed (approved):
No more issues from my side, great work!

aanm commented on tests/10-proxy.sh:
```diff
@@ -33,6 +33,12 @@ SVC_IP4="2.2.2.2"
 cleanup
 logs_clear
 
+function no_service_init {
```
@ianvernon needs restructure in ginkgo

ianvernon replied:
@jrajahalme can you file a ticket with the `area/CI` label to make sure that these test changes make it into the Ginkgo framework ? Thanks.

jrajahalme replied:
#2005

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -366,7 +433,6 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir string) error {
-			e.Mutex.Lock()
 			if createdPolicyMap {
 				// Remove policy map file only if it was created
 				// in this update cycle
```
the e.Consumable on the next line doesn't need to hold the lock()?

jrajahalme replied:
It does, but I think there is a possible race condition where the consumable WAS non-nil but was changed by another goroutine after we released the lock (i.e., endpoint's identity changed). To correctly implement the intended semantics (RemoveMap only if we added it, and to remove it from the same consumable we added it to) we need keep the original e.Consumable in a local variable and use that instead of e.Consumable at error case cleanup. Since we have a local reference now no locking is needed.

Thanks for pointing this out!

aanm replied:
Just to be sure, so the consumable itself doesn't need to have its lock held while changing its values?

jrajahalme replied:
Consumable's AddMap() and RemoveMap() methods take the lock on it, so they should be called with that lock not held.

Setting e.PolicyMap to nil in the error case seems dodgy, however. Endpoint's mutex is needed for that after all.

Thanks for digging deep enough on this.

jrajahalme replied:
Will squash in a fix for this reverting the removal of locking in this block & ask for one more review right after. Thanks!

jrajahalme replied:
OK, I squashed in the fix that reverts the removal of e.Mutex locking here. Not sure what I was thinking at the time I removed them, but at least we found a potential issue with the consumable being changed in between.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -1202,40 +1224,138 @@ func (e *Endpoint) GetDockerNetworkID() string {
-func (e *Endpoint) IsDisconnectingLocked() bool {
-	return e.State == StateDisconnected || e.State == StateDisconnecting
+// bumpPolicyRevision marks the endpoint to be running the next scheduled
+// policy revision as setup by e.regenerate(). endpoint.Mutex may not be held
```
s/  may not be held/ should not be held/

jrajahalme replied:
Changed, but really it "must not be held", as it will deadlock otherwise :-)

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -188,6 +188,7 @@ const (
 )
 
+// LabelsMap holds mapping from numeric policy identity to labels
 type LabelsMap map[policy.NumericIdentity]pkgLabels.LabelArray
```
`LabelsMap` can be private, no?

jrajahalme replied:
It could, but that would mess the nice naming of local variables (labelsMap). Also, after Thomas's KV-store improvements we can cache this centrally (maybe in policy/cache.go), so it would need to be public at that point. In short, I'd like to keep this public for now, too.

aanm commented on daemon/state.go:
```diff
@@ -97,21 +97,21 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
 			scopedLog := log.WithField(logfields.EndpointID, ep.ID)
-			if err := d.allocateIPs(ep); err != nil {
+			ep.Mutex.Lock()
+			if err := d.allocateIPsLocked(ep); err != nil {
```
Ideally we should move away from using mutexes outside the structures they protect

jrajahalme replied:
The intent here is to keep the change in endpoint and the corresponding endpoint.state change in the same critical section so that the state reported by e.g., "cilium endpoint list" is more reliable.

To move all the state changes to the endpoint package would require some more refactoring and exporting daemon's functions that are now private. It would be a good thing to do, but do you think this PR should be held for that, or if this could be addressed in a future PR?

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -293,6 +308,11 @@ type Endpoint struct {
 	// updated to and that will become effective with the next regenerate
 	nextPolicyRevision uint64
 
+	// forcePolicyCompute full endpoint policy recomputattion
```
recomputattion --> recomputation .

jrajahalme replied:
Fixed, thanks!

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -994,21 +1009,32 @@ func (e UpdateCompilationError) Error() string { return e.msg }
 	}
 
-	if opts != nil && !e.applyOptsLocked(opts) {
+	// Option changes may be overridden by the policy configuration.
```
I'm unclear as to what this means; can you please provide an example of this?

jrajahalme replied:
Endpoint's conntrack option may be disabled via CLI, but policy regeneration will revert that change back to enabled if the applicable policy requires conntrack (proxy redirect etc.).

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -1202,40 +1224,138 @@ func (e *Endpoint) GetDockerNetworkID() string {
+// SetStateLocked modifies the endpoint's state
+// endpoint.Mutex must be held
+// Returns true only if endpoints state was changed as requested
+func (e *Endpoint) SetStateLocked(toState string) bool {
```
I really like how the state machine logic is contained here, great job!

jrajahalme replied:
Thanks! I would have liked it even better if the conditions for the state transitions were here as well, but then the call sites should have needed extensive comments and would have been harder to understand.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -554,26 +589,28 @@ func (e *Endpoint) Regenerate(owner Owner) <-chan bool {
+	return changed, nil
 }
 
+// SetIdentity resets endpoints policy identity to 'id'.
```
endpoints --> endpoint's

jrajahalme replied:
Fixed, thanks!

ianvernon commented on tests/helpers.bash:
```diff
@@ -307,49 +307,31 @@ function wait_for_cilium_ep_gen {
+  local found
+  found=$(eval "$CMD")
+
+  while [[ "$found" -ne "$NUM_DESIRED" ]]; do
```
I'm quite glad to see this changed :-)

jrajahalme committed:
endpoint: Remove DeepCopy (dead code)

Endpoint's DeepCopy is not used any more, so it and the DeepCopy
functions for it's members can be removed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Remove dead code.

Policy cache iteration tracking was not used for anything, so the code
using it may be misleading. Better remove it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Minor clarifications.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
cilium: Make policy wait to wait for ready state.

cilium policy wait is used in tests to wait for all the endpoints to
have the given policy version. The intent is to be able to test that
policy right after, which is really possible only after all the
endpoints have also finished recomputing the new policy and
regenerated corresponding bpf.

This patch changes "cilium policy wait" to also wait for all the
endpoints to be in the "ready" state to facilitate the intended use
and make tests more deterministic, or "less flaky".

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Restructure proxy test.

Keep using the same docker containers for the whole test, only
changing the applied Cilium policy between tests. This both cuts
runtime (from 20 to 6 minutes) and provides a better stress test as
new policies are repeatedly applied to the same endpoints.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Pause after test failure if DEBUG is defined.

Debugging runtime test failures is easier if the containers and cilium state is not flushed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Use "NAT46" term consistently.

It was confusing when "nat64" and "nat46" were used
interchangeably. Change to consistently use "NAT46".

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Create redirects for TCP only.

Do not define parser if protocol is not TCP, so that we stop creating proxy redirects for UDP.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
api: Add endpoint state "waiting-to-regenerate"

This will be used in the next patch.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Do not error out on failed L4 policy removal.

Depending on the policy we may try installing the same L4 policy entry
to a bpf map multiple times, but only one of them get's actually
installed. On removal we'd try removing the same entry multiple times,
and only the first one actually succeeds. Remove the error condition
on failed removal to account for this case.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy/l3: remove "changed" flags

These flags were never cleared, so the code had to be able to deal
with failing removal of non-existent map anyway. Simplify by removing
the flags.

Removing state flags from policy structures also makes it easier to
reason about DeepEquals, which are to be used in subsequent patches.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Parse protocol to u8 in advance.

L4Policy Protocol values are already validated, so parsing them can't
fail.  Better yet, parse them in advance so there is a little bit less
work to do while endpoint is locked.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
tests: Streamline for more deterministic cilium endpoint state management.

Now that cilium tracks endpoint state more diligently, the tests can
rely on cilium status reports without second-quessing and looping to
double-check.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Do not take repo mutex when setting identity.

Policy repository is not used when changing the identity, so do not lock it.

Earlier we used Consumable's labels without locking, so taking a write
lock excluded policy regeneration while labels were changed. This is
no longer needed.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Refactor policy computation.

Refactor endpoint policy computation to:

1. Only recompute Consumable L4/7 policy if needed.
2. Compute new options before any bpf map updates, as some of them make a
   difference there.
3. Apply endpoint option changes only as part of policy regeneration, as
   options may be overridden depending on the applicable policy.
4. Fail out on failure before the endpoint policy has been touched in any
   way.
5. Add a new endpoint status "waiting-to-regenerate", which is entered
   as soon as we know endpoint policy needs to be recomputed. "ready"
   state is entered only when regeneration is done and there are no other
   pending changes that may require immediate policy recomputation.
6. State changes are made within the same critical section that gives
   rise to the state change. This reduces endpoint status churn exposed
   to the Cilium API.
7. Endpoint state changes are centralized into SetStateLocked() function
   to make reasoning about possible state transitions easier.

These steps allow endpoint policy to be recomputed in one pass, while
previously two passes were required. A second pass is still entered
when the bpf is regenerated, but most of it is skipped if nothing
changed since the previous pass.

Policy regeneration assumes endpoint's PolicyMap to exist, but it is
only created when the endpoint's bpf code is first generated. Bpf
regeneration also regenerates the policy, so we can safely skip the
initial policy regeneration.

In dry mode (used for unit testing) we must generate the policy
without generating bpf, hence we skip bpf updates in policy
regeneration in dry mode.

This fixes cilium-agent crash seen on k8s tests. To trigger this the
first pass through TriggerPolicyUpdatesLocked() needs to happen when:
- endpoint already has a security identity, and
- an L4 policy applies to the endpoint

Apparently this combination is not very common occurance, as we have
not hit this bug earlier.

In the long run we should shift all bpf changes to regenerateBPF() and
only recompute cilium-agent internal state in regeneratePolicy().

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
